kernel/eventfddev.c

```diff
@@ -0,0 +1,537 @@
+            ERAISE(-ENOMEM);
+
+        eventfd->magic = MAGIC;
+        eventfd->counter = initval;
+
+        if (flags & EFD_CLOEXEC)
+            eventfd->fdflags = FD_CLOEXEC;
+
+        eventfd->flags = flags;
+    }
+
+    *eventfd_out = eventfd;
+    eventfd = NULL;
+
+done:
+
+    if (eventfd)
+        free(eventfd);
+
+    return ret;
+}
+
+static ssize_t _read(
+    myst_eventfddev_t* eventfddev,
+    myst_eventfd_t* eventfd,
+    void* buf,
+    size_t count)
+{
+    ssize_t ret = 0;
+
+    if (!eventfddev || !_valid_eventfd(eventfd))
+        ERAISE(-EBADF);
+
+    if (!buf)
+        ERAISE(-EINVAL);
+
+    if (count < sizeof(uint64_t))
+        ERAISE(-EINVAL);
+
+    _lock(eventfd);
+
+    if (!_valid_eventfd(eventfd))
+    {
+        /* cannot unlock since mutex is no longer valid */
+        ERAISE(-EBADF);
+    }
+
+    /* if the counter is readable (non-zero) */
+    if (eventfd->counter != 0)
+    {
+        memcpy(buf, &eventfd->counter, sizeof(uint64_t));
+
+        if ((eventfd->flags & EFD_SEMAPHORE))
+            eventfd->counter--;
+        else
+            eventfd->counter = 0;
+
+        myst_cond_signal(&eventfd->cond);
+        _unlock(eventfd);
+        ret = sizeof(uint64_t);
+        goto done;
+    }
+
+    /* handle non-blocking read */
+    if (eventfd->flags & EFD_NONBLOCK)
+    {
+        _unlock(eventfd);
+        ERAISE(-EAGAIN);
+    }
+
+    /* wait here for another thread to write to the counter */
+    for (;;)
+    {
+        if (myst_cond_wait(&eventfd->cond, &eventfd->mutex) != 0)
+        {
+            /* unexpected */
+            _unlock(eventfd);
+            ERAISE(-EPIPE);
+        }
+
+        if (eventfd->counter != 0)
+        {
+            memcpy(buf, &eventfd->counter, sizeof(uint64_t));
+
+            if (eventfd->flags & EFD_SEMAPHORE)
+                eventfd->counter--;
+            else
+                eventfd->counter = 0;
+
+            myst_cond_signal(&eventfd->cond);
+            _unlock(eventfd);
+            ret = sizeof(uint64_t);
+            goto done;
+        }
+    }
+
+    /* unreachable */
+
+done:
+
+    if (ret > 0)
+        myst_tcall_poll_wake();
+
+    return ret;
+}
+
+static ssize_t _write(
+    myst_eventfddev_t* eventfddev,
+    myst_eventfd_t* eventfd,
+    const void* buf,
+    size_t count)
+{
+    ssize_t ret = 0;
+    uint64_t value;
+
+    if (!eventfddev || !_valid_eventfd(eventfd))
+        ERAISE(-EBADF);
+
+    if (!buf)
+        ERAISE(-EINVAL);
+
+    if (count < sizeof(uint64_t))
+        ERAISE(-EINVAL);
+
+    memcpy(&value, buf, sizeof(uint64_t));
+
+    if (value == 0xffffffffffffffff)
+    {
+        _unlock(eventfd);
+        ERAISE(-EINVAL);
+    }
+
+    _lock(eventfd);
+
+    if (!_valid_eventfd(eventfd))
+    {
+        /* cannot unlock since mutex is no longer valid */
+        ERAISE(-EBADF);
+    }
+
+    /* if able to write the the counter */
+    if (eventfd->counter + value < MAX_COUNTER_VALUE)
```

Comment:
The left side could wrap around (depending on the compiler). Suggest changing to `eventfd->counter < MAX_COUNTER_VALUE - value`.

Reply:
Nice idea. Fixed.